common/src/main/java/org/apache/celeborn/common/network/server/BufferStreamManager.java

```diff
@@ -313,74 +321,118 @@ public synchronized void setupDataPartitionReader(
               streamId,
               channel,
               () -> recycleStream(streamId));
-      // allocate resources when the first reader is registered
-      boolean allocateResources = readers.isEmpty();
-      readers.add(dataPartitionReader);
-      streamReaders.put(streamId, dataPartitionReader);
+      readers.put(streamId, dataPartitionReader);
 
-      // create initial buffers for read
-      if (allocateResources && buffers == null) {
+      // allocate resources when the first reader is registered
+      if (!bufferQueueInitialized) {
         memoryManager.requestReadBuffers(
-            minReadBuffers,
-            maxReadBuffers,
-            fileInfo.getBufferSize(),
-            (allocatedBuffers, throwable) ->
-                MapDataPartition.this.onBuffer(new LinkedBlockingDeque<>(allocatedBuffers)));
+            new ReadBufferRequest(
+                minReadBuffers,
+                maxReadBuffers,
+                fileInfo.getBufferSize(),
+                (allocatedBuffers, throwable) -> onBuffer(allocatedBuffers)));
+        bufferQueueInitialized = true;
       } else {
         triggerRead();
       }
     }
 
     // Read logic is executed on another thread.
-    public void onBuffer(Queue<ByteBuf> buffers) {
-      this.buffers = buffers;
+    public void onBuffer(List<ByteBuf> buffers) {
+      if (isReleased || bufferQueue.isReleased()) {
+        buffers.forEach(memoryManager::recycleReadBuffer);
+        return;
+      }
+
+      try {
+        bufferQueue.add(buffers);
+      } catch (Exception e) {
+        // this branch means that this bufferQueue is closed
+        buffers.forEach(memoryManager::recycleReadBuffer);
+        return;
+      }
+
       triggerRead();
+      isWaitingResources.set(false);
+      logger.debug(
+          "MapDataPartition set isWaitingResources to {} by on buffer", isWaitingResources.get());
     }
 
-    public void recycle(ByteBuf buffer, Queue<ByteBuf> bufferQueue) {
-      buffer.clear();
-      bufferQueue.add(buffer);
-      triggerRead();
+    public void recycle(ByteBuf buffer) {
+      bufferQueue.recycle(buffer);
+      if (isReleased || readers.isEmpty() || bufferQueue.isReleased()) {
+        return;
+      }
+
+      if (bufferQueue.size() > 0) {
+        triggerRead();
+      }
+
+      applyNewBuffers();
+    }
+
+    private void applyNewBuffers() {
```

Comment:
IMO this method is not logically correct. This method should do the following:
check whether bufferQueue.numBuffersOccupied is less than minReadBuffers, if so, request (min - numBuffersOccupied, max - numBuffersOccupied) buffers.
And this method should be checked by a thread, instead called in recycle, because recycle is multi-threaded.